File: java/src/test/java/com/company/automation/ApiTestBase.java
package com.company.automation;

import com.company.automation.clients.AuthClient;
import com.company.automation.clients.ProductsClient;
import com.company.automation.support.Config;
import com.company.automation.support.ConfigLoader;
import com.company.automation.support.extensions.MdcExtension;
import org.junit.jupiter.api.extension.ExtendWith;

/**
 * One thin base class for API and contract tests: clients and config, nothing else.
 *
 * <p>It stays thin on purpose. A base class that accumulates helpers becomes the place nobody can
 * safely change, and it hides where behaviour comes from. Cross-cutting behaviour belongs in a
 * JUnit extension, which is why logging context arrives via {@link MdcExtension} rather than a
 * {@code @BeforeEach} here.
 */
@ExtendWith(MdcExtension.class)
public abstract class ApiTestBase {

  protected final Config config = ConfigLoader.config();
  protected final ProductsClient products = new ProductsClient();
  protected final AuthClient auth = new AuthClient();

  /**
   * Registers disposable accounts. Any test that signs in uses one of these rather than a shared
   * account — see {@link com.company.automation.flows.AccountFlow} for what happened when they did
   * not.
   */
  protected final com.company.automation.flows.AccountFlow accounts =
      new com.company.automation.flows.AccountFlow();
}

File: csharp/tests/AutomationFramework.Tests/ApiTestBase.cs
using AutomationFramework.Core.Clients;
using AutomationFramework.Core.Flows;
using AutomationFramework.Core.Support;
using NUnit.Framework;

namespace AutomationFramework.Tests;

/// <summary>
/// One thin base class for API and contract tests: clients, config, and per-test log scope.
/// </summary>
/// <remarks>
/// It stays thin on purpose. A base class that accumulates helpers becomes the place nobody can safely
/// change, and it hides where behaviour comes from.
/// <para>
/// Clients are constructed per test instance, not shared statically. NUnit creates a fresh fixture
/// instance per test, so there is no shared mutable state to leak between them.
/// </para>
/// </remarks>
public abstract class ApiTestBase
{
    protected AppConfig Config { get; } = ConfigLoader.Config;

    protected ProductsClient Products { get; } = new(ApiHttp.Client);

    protected AuthClient Auth { get; } = new(ApiHttp.Client);

    protected UsersClient Users { get; } = new(ApiHttp.Client);

    /// <summary>
    /// Registers disposable accounts. Any test that signs in uses one of these rather than a shared account.
    /// </summary>
    protected AccountFlow Accounts { get; } = new(new UsersClient(ApiHttp.Client));

    [SetUp]
    public void BeginLogScope() => TestLog.BeginScope(TestContext.CurrentContext.Test.Name);

    [TearDown]
    public void EndLogScope() => TestLog.EndScope();
}
